BJS2-68795

tanufa25 opened the pull request:
Amazon warehouse task is done

Commit:
BJS2-68795
Amazon warehouse task is done

20cognosce commented on src/main/java/school/faang/bjs2_68795/amazon_warehouse/MessageFormatter.java:
```diff
@@ -0,0 +1,10 @@
+package school.faang.bjs2_68795.amazon_warehouse;
+
+public class MessageFormatter {
+    public static String format(String template, Object... args) {
+        for (Object arg : args) {
+            template = template.replaceFirst("\\{}", arg == null ? "null" : arg.toString());
+        }
```
Что это зачем

20cognosce replied:
По папочкам тогда раскладывай, что это к исключениям относится, имя детальнее описывай

20cognosce commented on src/main/java/school/faang/bjs2_68795/amazon_warehouse/Product.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.UUID;
+
+@ToString
+@EqualsAndHashCode
```
Зачем

20cognosce commented on src/main/java/school/faang/bjs2_68795/amazon_warehouse/Product.java:
```diff
@@ -0,0 +1,22 @@
+@EqualsAndHashCode
+@Getter
+public class Product {
+    private final UUID id;
```
Если у тебя итак есть уникальный идентификатор для объекта

20cognosce commented on src/main/java/school/faang/bjs2_68795/amazon_warehouse/ProductManager.java:
```diff
@@ -0,0 +1,76 @@
+
+    public void addProduct(Category category, String name) {
+        for (Product product : products) {
+            if (product.getName().equals(name) && product.getCategory().equals(category)) {
```
Objects.equals

20cognosce commented on src/main/java/school/faang/bjs2_68795/amazon_warehouse/ProductManager.java:
```diff
@@ -0,0 +1,76 @@
+    private static final String PRODUCT_NOT_FOUND = "Product with name {} and category {} is not found";
+    private static final String PRODUCT_ALREADY_EXISTS = "Product with name {} and category {} is already exists";
+
+    Set<Product> products = new HashSet<>();
```
где private final
никогда не забывайте блин ставить private final

это как выйти на улицу без штанов

20cognosce commented on src/main/java/school/faang/bjs2_68795/amazon_warehouse/ProductManager.java:
```diff
@@ -0,0 +1,76 @@
+    public void removeProduct(Category category, String name) {
+        boolean isFound = false;
+        for (Product product : products) {
+            if (product.getName().equals(name) && product.getCategory().equals(category)) {
```
Objects.equals

tanufa25 replied:
А у меня же только один объект, а второго нет. И у меня не получится создать второй такой же объект по имени и категории, так как в создании проверка на уникальность по совокупности этих значений. И даже если бы не было этой проверки, то когда я создам новый объект айдишники-то разные будут. В общем, не очень понимаю, как тут использовать  Objects.equals, если всё же его нужно использовать, то поясни, пожалуйста.

tanufa25 replied:
Разобралась, зачем он, и что так сравнивать безопаснее.

20cognosce commented on src/main/java/school/faang/bjs2_68795/amazon_warehouse/ProductManager.java:
```diff
@@ -0,0 +1,76 @@
+
+    public void removeProduct(Category category, String name) {
+        boolean isFound = false;
+        for (Product product : products) {
+            if (product.getName().equals(name) && product.getCategory().equals(category)) {
+                isFound = true;
+                products.remove(product);
+                break;
+            }
```
removeIf?
метод готовый есть

20cognosce commented on src/main/java/school/faang/bjs2_68795/amazon_warehouse/ProductManager.java:
```diff
@@ -0,0 +1,76 @@
+
+    public List<Product> findProductsByCategory(Category category) {
+        List<Product> categoryList = new ArrayList<>();
+        for (Product product : products) {
```
forEach

20cognosce commented on src/main/java/school/faang/bjs2_68795/amazon_warehouse/ProductManager.java:
```diff
@@ -0,0 +1,76 @@
+    public List<Product> findProductsByCategory(Category category) {
+        List<Product> categoryList = new ArrayList<>();
+        for (Product product : products) {
+            if (product.getCategory().equals(category)) {
```
Objects.equals

20cognosce commented on src/main/java/school/faang/bjs2_68795/amazon_warehouse/ProductManager.java:
```diff
@@ -0,0 +1,76 @@
+
+    public Map<Category, List<Product>> groupProductsByCategory() {
+        Map<Category, List<Product>> groupedProductsMap = new HashMap<>();
+        for (Product product : products) {
```
forEach

20cognosce commented on src/main/java/school/faang/bjs2_68795/amazon_warehouse/ProductManager.java:
```diff
@@ -0,0 +1,76 @@
+
+    public void printAllProducts() {
+        Map<Category, List<Product>> groupProductsByCategoryMap = groupProductsByCategory();
+        for (Map.Entry<Category, List<Product>> entry : groupProductsByCategoryMap.entrySet()) {
+            log.info("Category: {}", entry.getKey());
+            log.info("Products:");
+            for (Product product : entry.getValue()) {
```
forEach

Commit:
BJS2-68795
MessageFormatter is transformed to ExceptionMessageFormatter and added to exception package with 2 exception classes.
EqualsAndHashCode annotation is removed as there is unique UUID.
Objects.equals are used instead of obj1.equals(obj2)
private final modificators for Set<Product> are added.
RemoveIf method for Set<Product> is used.
ForEach method is used instead of for.

20cognosce commented on src/main/java/school/faang/bjs2_68795/amazon_warehouse/ProductManager.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    public void removeProduct(Category category, String name) {
+        if (!products.removeIf(
+                product -> Objects.equals(product.getName(), name)
+                        && Objects.equals(product.getCategory(), category))) {
```
не выполнять модификации внутри if
проверка не должна изменять ничье состояние

20cognosce commented on src/main/java/school/faang/bjs2_68795/amazon_warehouse/ProductManager.java:
```diff
@@ -0,0 +1,71 @@
+
+    public void printAllProducts() {
+        Map<Category, List<Product>> groupProductsByCategoryMap = groupProductsByCategory();
+        for (Map.Entry<Category, List<Product>> entry : groupProductsByCategoryMap.entrySet()) {
```
почему не forEach

20cognosce reviewed (approved):
Принято
